### DISCUSSION TITLE
Query Regarding Array Bound Checks and Error Handling in MOOSE

### DISCUSSION THREAD

**Discussion by shrituntunroy** [idaholab/moose#28463] [Q&A Modules: Solid mechanics] [answered]
Hi Moose Team,

I am a new MOOSE user and currently learning to use the software.

Before switching to MOOSE, I was working with Abaqus where I often used Fortran subroutines (like UMAT, UEXTERNALDB, UEL, etc.) to implement programmable features.

In Fortran subroutine, I would define arrays as follows:

```fortran
real(8) :: slip_dir(num_slip_sys)
real(8) :: slip_nor(num_slip_sys)
```

There, If I accidentally accessed an array element beyond its defined range, the subroutine would either fail to compile or throw a segmentation fault during execution, making it easy to identify the issue and correct it.

In MOOSE, the common practice for defining arrays is:

```c++
std::vector<Real> slip_dir(_num_slip_sys); 
std::vector<Real> slip_nor(_num_slip_sys);
```

However, here,  if I mistakenly access an out-of-bounds index in MOOSE, I don’t get a compile-time or segmentation fault error. Instead, I typically encounter issues like poor convergence, floating-point exceptions, or other runtime errors, but not a clear segmentation fault pointing to the mistake.

My question is: Are there any ways to define arrays, or any specific compiler settings, that would catch out-of-bounds array access during compilation or execution, resulting in a more informative error (like a segmentation fault)?

Your guidance on this would be greatly appreciated.

Thank you for your support!

**Comment by GiudGiud** [comment]
Hello

If you compile in debug mode (set the environment variable METHOD=dbg ) I have observed it will complain when you access out of bounds. I'm not sure how what compiler option turned that on. 

An alternative is to add mooseAssert in your code to check that you do not access out of bound by comparing the index and the size
The assert is only compiled in devel and debug mode so it does not affect the opt mode performance

**Reply by GiudGiud** [reply]
This won't catch anything at compilation btw. Only execution

**Reply by GiudGiud** [reply]
This won't catch anything at compilation btw. Only execution

**Reply by shrituntunroy** [reply]
Hi  Giudicelli,

Thankyou for your reply.

Can you give an example on how to add mooseAssert for this case. I looked up in Moose Source files but didn't found any relevant case.

**Reply by GiudGiud** [reply, marked as answer]
For example
```
mooseAssert(index<vector.size(), "Accessing out of bounds");
```
if the condition in the first part of the call is false, then the second argument, a string, is output as an error message

You can use this before indexing into the vector with vector[index]

**Reply by shrituntunroy** [reply]
Thankyou  Giudicelli,

for your help
